### PR TITLE
lookup: fall back to a slow path (list all + check) when necessary

### PR DESCRIPTION
**Pull request by ecordell**
Fixes https://github.com/authzed/spicedb/issues/470

### CODE REVIEW

File: internal/dispatch/graph/graph.go [hunk 1]
```diff
@@ -184,6 +186,46 @@ func (ld *localDispatcher) DispatchLookup(ctx context.Context, req *v1.DispatchL
 		Revision:              revision,
 	}
 
+	relation, err := ld.loadRelation(ctx, req.ObjectRelation.Namespace, req.ObjectRelation.Relation, revision)
+	if err != nil {
+		return &v1.DispatchLookupResponse{Metadata: emptyMetadata, ResolvedOnrs: []*v0.ObjectAndRelation{}}, nil
+	}
+
+	// check top-level rewrite
+	slowPath := false
```
Comment (josephschorr) [inline internal/dispatch/graph/graph.go:195]: Extract into `requiresSlowPath` function?

File: internal/graph/lookup.go [hunk 2]
```diff
@@ -69,6 +73,98 @@ func (cl *ConcurrentLookup) Lookup(ctx context.Context, req ValidatedLookupReque
...
+		objects.Add(&v0.ObjectAndRelation{
+			Namespace: tpl.ObjectAndRelation.Namespace,
+			ObjectId:  tpl.ObjectAndRelation.ObjectId,
+			Relation:  req.ObjectRelation.Relation,
+		})
+	}
+
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
+	g, checkCtx := errgroup.WithContext(cancelCtx)
```
Comment (josephschorr) [inline internal/graph/lookup.go:106]: newline before this line

File: internal/graph/lookup.go [hunk 3]
```diff
@@ -69,6 +73,115 @@ func (cl *ConcurrentLookup) Lookup(ctx context.Context, req ValidatedLookupReque
...
+
+	g, checkCtx := errgroup.WithContext(cancelCtx)
+
+	type onrWithMeta struct {
+		onr  *v0.ObjectAndRelation
+		meta *v1.ResponseMeta
+	}
+	allowedONRs := make(chan onrWithMeta)
+
+	sem := make(chan struct{}, MaxConcurrentSlowLookupChecks)
```
Comment (jakedt) [inline internal/graph/lookup.go:115]: Can we add some metrics so we know how often this is happening, and what proportion of them are succeeding vs failing?

Comment (jzelinskie) [inline internal/graph/lookup.go:115]: This seems like a good place to use https://pkg.go.dev/golang.org/x/sync/semaphore and not just a channel.

File: internal/graph/lookup.go [hunk 4]
```diff
@@ -69,6 +73,115 @@ func (cl *ConcurrentLookup) Lookup(ctx context.Context, req ValidatedLookupReque
...
+	if it.Err() != nil {
+		resp := lookupResultError(req, err, emptyMetadata)
+		return resp.Resp, resp.Err
+	}
+
+	var groupErr error
+	go func() {
+		groupErr = g.Wait()
+		close(allowedONRs)
+	}()
```
[the remark this answers is not in the document]
Comment (ecordell) [reply internal/graph/lookup.go:153]: closing `allowedONRs` is the stop condition for the loop that reads from `allowedONRs`, and I want to be reading from the channel while writes are still happening

[the remark this answers is not in the document]
Comment (ecordell) [reply internal/graph/lookup.go]: I came around on this and put the consumer in a goroutine in the errgroup as well

File: internal/dispatch/graph/graph.go [hunk 5]
```diff
@@ -183,10 +191,55 @@ func (ld *localDispatcher) DispatchLookup(ctx context.Context, req *v1.DispatchL
...
+		return ld.lookupHandler.LookupViaChecks(ctx, validatedReq)
+	}
 
 	return ld.lookupHandler.Lookup(ctx, validatedReq)
 }
 
+func (ld *localDispatcher) requiresLookupViaChecks(relation *v0.Relation) bool {
+	// TODO: refactor walker so that we don't have to make two separate checks
+	// check top-level rewrite
+	slowPath := false
```
Comment (jzelinskie) [inline internal/dispatch/graph/graph.go:211]: We can probably just return early instead of using a var

File: internal/graph/lookup.go [hunk 6]
```diff
@@ -69,6 +74,116 @@ func (cl *ConcurrentLookup) Lookup(ctx context.Context, req ValidatedLookupReque
 	return resolved.Resp, resolved.Err
 }
 
+// LookupViaChecks performs a slow-path lookup request with the provided request and context.
+// it performs a lookup by finding all instances of the requested object and then
```
Comment (jzelinskie) [inline internal/graph/lookup.go:78]: capitalize "it"

File: internal/dispatch/graph/graph.go [hunk 7]
```diff
@@ -183,10 +191,55 @@ func (ld *localDispatcher) DispatchLookup(ctx context.Context, req *v1.DispatchL
...
+	rw := relation.GetUsersetRewrite()
+	if rw != nil {
```
Comment (jzelinskie) [inline internal/dispatch/graph/graph.go:213]: this can be done in one line

**Review by jzelinskie — APPROVED** [review]

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**